### DISCUSSION TITLE
Post results to Error 405 Method not Allowed

### DISCUSSION THREAD

**Discussion by Finix21** [fastapi/fastapi#7859] [Questions] [answered]
Used the very code on the docs for simple posting and it only allows me to use get

**Comment by Finix21** [comment]
actually the log shows im using GET although in code im using POST

**Comment by nikhilno1** [comment]
What was the resolution for this issue? I am getting same behavior.

**Reply by LucasColas** [reply]
Did you find a solution ?

**Reply by CarlosOliveira-23** [reply]
Prezados Bom Dia, 
Também estou com o mesmo problema.

**Comment by tiangolo** [comment]
Thanks @Finix21 for reporting back and closing the issue :+1: 

@nikhilno1 you might want to follow the tutorial exactly as is. With the integrated docs UI to send the POST and confirm that it works. Then you can check in your own code, gradually what is different.

**Comment by BramVanroy** [comment]
@tiangolo  Hi. I am running into the same issue as mentioned above. Going through the tutorial, and not one of [the snippets](https://fastapi.tiangolo.com/tutorial/body/) here work. The issue is, as @Finix21 mentions, that the code correctly states that the method should run on a POST request, the server executes it as a GET request.

As an example, this code:

```python
from typing import Optional

from fastapi import FastAPI
from pydantic import BaseModel


class Item(BaseModel):
    name: str
    description: Optional[str] = None
    price: float
    tax: Optional[float] = None


app = FastAPI()


@app.post("/items/")
async def create_item(item: Item):
    return item
```

will result in `{"detail":"Method Not Allowed"}`, and the terminal reports:

```
←[32mINFO←[0m:     127.0.0.1:57638 - "←[1mGET /items/ HTTP/1.1←[0m" ←[31m405 Method Not Allowed←[0m
```

This is true for all snippets on that page. I have tried without `--reload`, tried restarting the server.

Running on Windows 10 64x, Python 3.8.2:
 - fastapi    0.61.1
 - starlette  0.13.6
 - uvicorn    0.12.2

**Comment by ycd** [comment]
@BramVanroy can you try sending a curl request? The code snippet you have works well for me

```
curl -X POST "http://127.0.0.1:8000/items/" -H  "accept: application/json" -H  "Content-Type: application/json" -d "{\"name\":\"string\",\"description\":\"string\",\"price\":0,\"tax\":0}"
```

```python
Out: {"name":"string","description":"string","price":0.0,"tax":0.0}
```

**Comment by BramVanroy** [comment]
@ycd The cURL request works, but the browser doesn't seem to be playing nicely:

```
←[32mINFO←[0m:     127.0.0.1:65023 - "←[1mGET /items/ HTTP/1.1←[0m" ←[31m405 Method Not Allowed←[0m
```

I am just surfing to `http://127.0.0.1:8000/items/` and the server instance throws the error above. I also tried `http://127.0.0.1:8000/items/?name=potatoes&price=0.5` instead, but to no avail. I also tried different browsers (as this seems browser related rather than server related), but no luck.

**Comment by BramVanroy** [comment]
So the issue was that the example of course will not run because we never do a POST operation. Perhaps that can be improved in the documentation, as inexperienced users may think that all code snippets are executable - but the example above will not work (because we are just GETting).

**Comment by ycd** [comment]
The error is raised by HTTP server it never reaches to FastAPI you can check that by using `Request.method`, [since `GET` method can not contain a body](https://swagger.io/docs/specification/describing-request-body/), , I think the HTTP server assumes it is a `GET` method and returns 405

**Comment by sanpj2292** [comment]
When we send the request as mentioned in the image below
![image](https://user-images.githubusercontent.com/18283111/112417372-7df2e280-8d4d-11eb-87a8-4dcb2b2646b3.png)

```python
@app.put('/items/{item_id}/')
async def put_items(item_id: int, importance : int = Body(...)):
    return {
        'item_id': item_id,
        'importance': importance,
    }
```

We get an appropriate response. I think this is a bug.

**Comment by dnhyde** [comment]
I am having the same issue. On localhost everything works as expected from postman. Then after deploying it on google cloud run I keep having `405 Method Not Allowed` I could not understand what was the issue in this thread, and I cannot understand if this is related to google cloud run and why, any hints? 🙏

**Reply by wangxiaobai-123** [reply]
i have the same problem exactly.postman runs well ,but i use uni.request() to send post request to endpoint ,error happened.

**Comment by caioalmeida97** [comment]
@dnhyde I'm having the exact same problem as you. I deployed my FastAPI source code to Cloud Run and now when I try to make a POST, it returns `405 Method Not Allowed` as well. Did you solve your problem? If so, how did you do it?

**Comment by caioalmeida97** [comment]
@dnhyde I think I solved my issue. I was trying to make a POST request to my Google Cloud Run service without `https://` prefix and I was getting the status code `405`. I just added the HTTPS prefix and it worked.

**Comment by JulienBallaudOptima** [comment]
@caioalmeida97 : thank you SO MUCH for this "HTTPS prefix" tip, it really helped me to solve the same problem I was facing 👍👍

**Comment by sparshgarg23** [comment]
I tried adding https prefix but I ended up getting  this site can't be reached
in the log  the error is invalid https request recieved.
Would appreciate if someone can guide me and other people how to upload files so that we can later on use it to make predictions

**Comment by gauravgund** [comment]
I am still getting error 405. Could anyone tell how to solve it?

**Comment by mechkit** [comment, marked as answer]
For me the problem was that because of a config mismatch when converting from Node, the route was at 
> http://...ip... /custom_route/

instead of
> http:// ...ip... /custom_route

and changing it fixed the problem.

**Comment by rbracco** [comment]
For me it was also the lack of a trailing /, using requests.post to a post route. It would be great if this could be fixed to raise another error because this was a tricky debug.

**Reply by SourenGhosh** [reply]
Exactly, i have wasted the whole day for this trailing /.

**Reply by rainer85ah** [reply]
me too :(

**Comment by ghost** [comment]
i had the same issue because i was mounting `StaticFiles` on the root path (`/`), meaning all my POSTs were interpreted as static file requests

**Reply by netsesame** [reply]
I had the same problem as you. How did you fix it?

**Reply by gagank1** [reply]
Mount StaticFiles to /static or something similar. If you're using react, change your project's home directory to fix the linking at build time: https://stackoverflow.com/questions/44115029/create-react-app-npm-run-build-issues

**Comment by tiangolo** [comment]
Hey all! Thanks for the conversation here. If anyone is still having problems, it's better to create a new issue, following the template and writing a self contained example. If you check closely, even though the problem was visible as a 405 error, the cause was very different in several cases, so, it was actually not the same problem. That's why it's better to create a new issue with all the description and a self contained example.

**Comment by yabenz** [comment]
Hi, I'm trying to use the fast api with Stable Difussion UI -- the Automatic 1111
I installed Automatic1111 the webui works on [gradio.app](http://gradio.app/)


i tried to follow this : https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/API
but I get the error:
---------------------------------------------------------------------------
KeyError Traceback (most recent call last)
Cell In[24], line 35
31 r = response.json()
32 print("Got response")
---> 35 for i in r['images']:
36 image = [Image.open](http://image.open/)(io.BytesIO(base64.b64decode(i.split(",",1)[0])))
38 png_payload = {
39 "image": "data:image/png;base64," + i
40 }

KeyError: 'images'

**Comment by LR1093141031** [comment]
因为我挂载StaticFiles在根路径 ( /) 上，这意味着我所有的 POST 都被解释为静态文件请求

**Comment by agalbachicar** [comment]
I found this issue as well. Please find here a great explanation for those cases in which serving static html files might be colliding with depth routes: https://stackoverflow.com/a/73113792

**Comment by rainer85ah** [comment]
Hi Lucas,

 I did. A novice mistake due to the lack of examples and guideline out
there. With API usually you will find a lot of fetch examples online but
not post examples. So my post api method had the wrong argument type.

Hope this helps.
Rainer

**Reply by LucasColas** [reply]
But how did you figure it out ? I'm also stuck with the same error.

**Comment by rainer85ah** [comment]
I spent days troubleshooting my frontend post function, I wasn't sure my
javascript function was correct as a JS newbie, and also troubleshooting my
nginx (web server and load balancer) in case it was changing the type of
request or redirecting incorrectly, so after trying everything that made
sense to me at the time, I decided to start looking elsewhere so I
arrive to the API from there I tried different version of the function,
play with different types of possible annotations, etc... until I started
playing with the arguments of the post function.

This is the working method:
from fastapi import Body

@router.post(path="/api/endpoint/add_email")
def add_email(email: str = Body()) -> bool:
      pass

The bad example:
@router.post(path="/api/endpoint/add_email")
def add_email(email: str = ***@***.***") -> bool:
      pass

In conclusion, my FastApi wasn't assigning the request to my add_email
endpoint due to the argument type or the endpoint itself didn't understand
the argument I was passing to it.

Again, hope this helps.
Rainer

**Reply by LucasColas** [reply]
I still don't understand what makes your code working.
I'm trying to do a code with post (to upload images) : 
```py
from fastapi import FastAPI, UploadFile, File
import io
from PIL import Image
from typing import Union


def perform_object_detection(image):
    pass

app = FastAPI()

@app.post("/detect")
async def detect_objects(file: UploadFile = File(...)):
    contents = await file.read()
    image = Image.open(io.BytesIO(contents))

    results = perform_object_detection(image)

    return results

```

**Reply by rainer85ah** [reply]
@LucasColas Sorry buddy I haven't use post endpoint for images, your code look good to me but probably it will need something different like:

- The annotation: My code use router because I create the apirouter object this way: router = APIRouter()
- The Path: My code use the full path of the API endpoint "/api/endpoint/add_email", maybe you need to add something like this "/api/detect"
- Then, I will look into other type of arguments in case there is, instead of UploadFile.
- Finally, Maybe is the File(...) part, maybe is expecting one image and you are uploading multiples images or the other way around.

Good luck!!!

**Reply by LucasColas** [reply]
No I don't upload multiple images nor one image. I just have : 
`{"detail":"Method Not Allowed"}`

**Comment by r-moctezuma** [comment]
For people who are experiencing this - this is what I found in my case. Everything worked locally, but we had weird behavior when deploying to Lightsail. Things like HTTP to HTTPS issues, or routes getting mixed up. The root cause was deploying FastAPI behind a proxy/load balancer (Lightsail's) that gets confused. In my case, the --proxy-headers parameter for uvicorn did the trick.

**Reply by silentrob** [reply]
This worked for me. Thanks

**Comment by liuxuan1-1** [comment]
I encountered this problem too. It worked well locally, but when I deployed it to the Azure App Service, I got this error: 405 Method Not Allowed. After a day and a half, I finally found the root cause. It was just missing an 's'; I should have used HTTPS instead of HTTP.

**Comment by SuryaaSeran** [comment]
Change http to https for POST request and it shall work.
